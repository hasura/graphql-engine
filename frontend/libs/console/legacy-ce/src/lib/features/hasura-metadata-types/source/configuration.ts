export type FromEnv = { from_env: string };
type ValidJson = Record<string, any>;

export interface PostgresConfiguration {
  connection_info: {
    database_url:
      | string
      | FromEnv
      | {
          username: string;
          password?: string;
          database: string;
          host: string;
          port: string;
        };
    pool_settings?: {
      max_connections?: number;
      idle_timeout?: number;
      retries?: number;
      pool_timeout?: number;
      connection_lifetime?: number;
    };
    use_prepared_statements?: boolean;
    /**
     * The transaction isolation level in which the queries made to the source will be run with (default: read-committed).
     */
    isolation_level?: 'read-committed' | 'repeatable-read' | 'serializable';
    ssl_configuration?: {
      sslmode: string;
      sslrootcert: FromEnv;
      sslcert: FromEnv;
      sslkey: FromEnv;
      sslpassword: FromEnv;
    };
  };
  /**
   * Optional list of read replica configuration (supported only in cloud/enterprise versions)
   */
  read_replicas?: PostgresConfiguration['connection_info'][];
  /**
   * Name of the schema where the graphql-engine will install database extensions (default: public)
   */
  extensions_schema?: any;
}

export interface MssqlConfiguration {
  connection_info: {
    connection_string: string | FromEnv;
    pool_settings?: {
      max_connections?: number;
      idle_timeout?: number;
    };
  };
  read_replicas?: MssqlConfiguration['connection_info'][];
}

export interface BigQueryConfiguration {
  service_account: string | ValidJson | FromEnv;
  project_id: string | FromEnv;
  datasets: string[] | FromEnv;
}

export interface CitusConfiguration extends PostgresConfiguration {}
