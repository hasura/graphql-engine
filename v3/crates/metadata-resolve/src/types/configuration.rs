/// Configuration for the metadata-resolve step.
///
/// Deserialization is only intended to be used for testing and is not reliable.
#[derive(Debug, Clone, Default, serde::Deserialize)]
#[serde(default, deny_unknown_fields, rename_all = "camelCase")]
pub struct Configuration {
    pub allow_unknown_subgraphs: bool,
    pub unstable_features: UnstableFeatures,
}

/// internal feature flags used in metadata resolve steps
///
/// Deserialization is only intended to be used for testing and is not reliable.
#[derive(Debug, Clone, Copy, Default, serde::Deserialize)]
#[serde(default, deny_unknown_fields, rename_all = "camelCase")]
#[allow(clippy::struct_excessive_bools)]
pub struct UnstableFeatures {
    pub enable_order_by_expressions: bool,
    pub enable_ndc_v02_support: bool,
    pub enable_subscriptions: bool,
    pub enable_jsonapi: bool,
}
