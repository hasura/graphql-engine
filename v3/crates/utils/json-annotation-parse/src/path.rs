use crate::types::Value;
use jsonpath::{JSONPath, JSONPathElement};

#[derive(Debug)]
pub enum WalkError<Ann> {
    PathValueMismatch(JSONPathElement, Value<Ann>),
    MissingIndex(usize, Value<Ann>),
    MissingKey(String, Value<Ann>),
}

// reverse the path then walk it
pub fn walk<'a, Ann: Clone>(
    value: &'a Value<Ann>,
    path: &mut JSONPath,
) -> Result<&'a Ann, WalkError<Ann>> {
    let mut reversed_path = JSONPath(path.0.iter().rev().cloned().collect());
    walk_inner(value, &mut reversed_path)
}

fn walk_inner<'a, Ann: Clone>(
    value: &'a Value<Ann>,
    path: &mut JSONPath,
) -> Result<&'a Ann, WalkError<Ann>> {
    match path.0.pop() {
        Some(tail) => {
            match (value, &tail) {
                (Value::Array(_, items), JSONPathElement::Index(i)) => {
                    match items.get(*i) {
                        Some(item) => {
                            // got it!
                            walk_inner(item, path)
                        }
                        None => Err(WalkError::MissingIndex(*i, value.clone())),
                    }
                }
                (Value::Object(_, items), JSONPathElement::Key(s)) => {
                    match items.get(s) {
                        Some(item) => {
                            // got it!
                            walk_inner(item, path)
                        }
                        None => Err(WalkError::MissingKey(s.to_string(), value.clone())),
                    }
                }
                _ => Err(WalkError::PathValueMismatch(tail.clone(), value.clone())),
            }
        }
        None => {
            // we are here, return outer annotation
            match value {
                Value::Bool(ann, _) => Ok(ann),
                Value::Null(ann) => Ok(ann),
                Value::Number(ann, _) => Ok(ann),
                Value::String(ann, _) => Ok(ann),
                Value::Array(ann, _) => Ok(ann),
                Value::Object(ann, _) => Ok(ann),
            }
        }
    }
}
