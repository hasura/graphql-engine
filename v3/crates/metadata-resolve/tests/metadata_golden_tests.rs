//! Tests that attempt to resolve different metadata files and assert that they parse successfully
//! or fail in the expected way.

use std::fs;
use std::path::Path;

use metadata_resolve::configuration;

#[test]
fn test_passing_metadata() {
    insta::glob!("passing/**/metadata.json", |path| {
        let directory = path.parent().unwrap();
        insta::with_settings!({
            snapshot_path => directory,
            snapshot_suffix => "",
            prepend_module_to_snapshot => false,
        }, {
            let configuration = read_test_configuration(directory)
                .unwrap_or_else(|error| panic!("{}: Could not read configuration: {error}",directory.display()));

            let metadata_json_text = std::fs::read_to_string(path)
                .unwrap_or_else(|error| panic!("{}: Could not read file {path:?}: {error}", directory.display()));

            let metadata_json_value = serde_json::from_str(&metadata_json_text)
                .unwrap_or_else(|error| panic!("{}: Could not parse JSON: {error}",directory.display()));

            let metadata = open_dds::traits::OpenDd::deserialize(metadata_json_value)
                .unwrap_or_else(|error| panic!("{}: Could not deserialize metadata: {error}", directory.display()));

            let resolved = metadata_resolve::resolve(metadata, &configuration)
                .unwrap_or_else(|error| panic!("{}: Could not resolve metadata: {error}",directory.display()));

            insta::assert_debug_snapshot!("resolved", resolved);
        });
    });
}

#[test]
fn test_failing_metadata() {
    insta::glob!("failing/**/metadata.json", |path| {
        let directory = path.parent().unwrap();
        insta::with_settings!({
            snapshot_path => directory,
            snapshot_suffix => "",
            prepend_module_to_snapshot => false,
        }, {
            let configuration = read_test_configuration(directory)
                .unwrap_or_else(|error| panic!("{}: Could not read configuration: {error}", directory.display()));

            let metadata_json_text = std::fs::read_to_string(path)
                .unwrap_or_else(|error| panic!("{}: Could not read file {path:?}: {error}", directory.display()));

            match serde_json::from_str(&metadata_json_text) {
                Ok(metadata_json_value) => {
                    match open_dds::traits::OpenDd::deserialize(metadata_json_value) {
                        Ok(metadata) => {
                            match metadata_resolve::resolve(metadata, &configuration) {
                                Ok(_) => {
                                    panic!("{}: Unexpected success when resolving {path:?}.", directory.display());
                                }
                                Err(msg) => {
                                    insta::assert_snapshot!("resolve_error", msg);
                                }
                            }
                        }
                        Err(msg) => {
                            insta::assert_snapshot!("deserialize_error", msg);
                        }
                    };
                }

                Err(msg) => {
                    insta::assert_snapshot!("parse_error", msg);
                }
            };
        });
    });
}

fn read_test_configuration(
    directory: &Path,
) -> Result<configuration::Configuration, Box<dyn std::error::Error>> {
    let unstable_features = configuration::UnstableFeatures {
        enable_order_by_expressions: false,
        enable_ndc_v02_support: false,
        enable_subscriptions: false,
        enable_jsonapi: false,
    };

    let configuration_path = directory.join("configuration.json");
    if configuration_path.exists() {
        let reader = fs::File::open(configuration_path)?;
        let configuration = serde_json::from_reader(reader)?;
        Ok(configuration::Configuration {
            unstable_features,
            ..configuration
        })
    } else {
        Ok(configuration::Configuration {
            allow_unknown_subgraphs: false,
            unstable_features,
        })
    }
}
