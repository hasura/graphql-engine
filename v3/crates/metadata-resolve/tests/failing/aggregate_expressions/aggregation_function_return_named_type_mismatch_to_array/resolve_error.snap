---
source: crates/metadata-resolve/tests/metadata_golden_tests.rs
expression: string
input_file: crates/metadata-resolve/tests/failing/aggregate_expressions/aggregation_function_return_named_type_mismatch_to_array/metadata.json
---
Error: the aggregate expression Int4_aggregate_exp (in subgraph default) specifies an aggregation function '_top5' which is mapped to the data connector 'mypg (in subgraph default)' but the Open DD return type Int4 (in subgraph default) is not compatible with the data connector's return type. Reason: The data connector's return type is an array, but the Open DD return type is not
